pyo3-macros-backend/src/pyproto.rs

```diff
@@ -129,49 +130,61 @@ fn submit_normal_methods(py_methods: Vec<TokenStream>, ty: &syn::Type) -> TokenS
-    if proto.extension_trait == "" {
-        return Ok(quote! {});
+) -> TokenStream {
+    if proto.slots_trait.is_empty() {
+        return TokenStream::default();
     }
-    let ext_trait: syn::Path = syn::parse_str(proto.extension_trait)?;
-    let mut tokens = vec![];
+
+    let module = proto.module();
+    let slots_trait = syn::Ident::new(proto.slots_trait, Span::call_site());
+    let slots_trait_slots = syn::Ident::new(proto.slots_trait_slots, Span::call_site());
+
     if proto.name == "Buffer" {
-        // For buffer, we construct `PyProtoMethods` from PyBufferProcs
-        tokens.push(quote! {
-            let mut proto_methods = pyo3::ffi::PyBufferProcs::default();
-        });
-        for getter in proto.slot_getters(method_names) {
-            let get = syn::Ident::new(getter, Span::call_site());
-            let field = syn::Ident::new(&format!("bf_{}", &getter[4..]), Span::call_site());
-            tokens.push(quote! { proto_methods.#field = Some(<#ty as #ext_trait>::#get()); });
-        }
-    } else {
-        // For other protocols, we construct `PyProtoMethods` from Vec<ffi::PyType_Slot>
-        tokens.push(quote! { let mut proto_methods = vec![]; });
-        for getter in proto.slot_getters(method_names) {
-            let get = syn::Ident::new(getter, Span::call_site());
-            tokens.push(quote! {
-                let slot = <#ty as #ext_trait>::#get();
-                proto_methods.push(pyo3::ffi::PyType_Slot { slot: slot.0, pfunc: slot.1 as _ });
-            });
-        }
-    };
-    if tokens.len() <= 1 {
-        return Ok(quote! {});
+        return quote! {
+            impl pyo3::class::proto_methods::#slots_trait<#ty>
+            for pyo3::class::proto_methods::PyClassProtocols<#ty> {
+                fn #slots_trait_slots(
+                    self
+                ) -> Option<&'static pyo3::class::proto_methods::PyBufferProcs> {
+                    static PROCS: pyo3::class::proto_methods::PyBufferProcs
+                        = pyo3::class::proto_methods::PyBufferProcs {
+                            bf_getbuffer: Some(pyo3::class::buffer::getbuffer::<#ty>),
+                            bf_releasebuffer: Some(pyo3::class::buffer::releasebuffer::<#ty>),
+                        };
+                    Some(&PROCS)
+                }
+            }
+        };
     }
-    Ok(quote! {
-        pyo3::inventory::submit! {
-            #![crate = pyo3] {
-                type Inventory =
-                    <#ty as pyo3::class::proto_methods::HasProtoInventory>::ProtoMethods;
-                <Inventory as pyo3::class::proto_methods::PyProtoInventory>::new(
-                    { #(#tokens)* proto_methods.into() }
-                )
+
+    let mut tokens = proto
+        .slot_defs(method_names)
+        .map(|def| {
+            let slot = syn::Ident::new(def.slot, Span::call_site());
+            let slot_impl: syn::Path = syn::parse_str(def.slot_impl).unwrap();
```

Comment:
I can understand it merely panics, but returning a `Result` would be more kind for future developers.

Reply:
Ah actually since the refactoring to `defs.rs` this only needs `syn::Ident` here, so I can remove the panic!